packages/keystone/runtimeStats.js

```diff
@@ -0,0 +1,171 @@
+                const xTargetHeader = req.headers['x-target']
+
+                const xTarget = possibleXTargetValues.includes(xTargetHeader) ? xTargetHeader : 'other'
+
+                const isPost = method === 'post'
+
+                const isAdminApiUrl = url === '/admin/api'
+                const isApiUrl = url.startsWith('/api')
+                const isOidcUrl = url.startsWith('/oidc')
+                const isWellKnownUrl = url.startsWith('/.well-known')
+                const isHealthCheckUrl = url.startsWith(DEFAULT_HEALTHCHECK_URL)
+
+                /** @type {'api' | 'graphql' | 'oidc' | 'wellKnown' | 'healthCheck' | 'other'} */
+                let requestType = 'other'
+
+                if (isPost && isAdminApiUrl) {
+                    requestType = 'graphql'
+                } else if (isApiUrl) {
+                    requestType = 'api'
+                } else if (isOidcUrl) {
+                    requestType = 'oidc'
+                } else if (isWellKnownUrl) {
+                    requestType = 'wellKnown'
+                } else if (isHealthCheckUrl) {
+                    requestType = 'healthCheck'
+                }
+
+                runtimeStats.activeRequestsIds.add(req.id)
+
+                runtimeStats.activeRequestsCountByType[requestType] = (runtimeStats.activeRequestsCountByType[requestType] || 0) + 1
+                runtimeStats.activeRequestsCountByTarget[xTarget] = (runtimeStats.activeRequestsCountByTarget[xTarget] || 0) + 1
+
+                runtimeStats.totalRequestsCount = (runtimeStats.totalRequestsCount || 0) + 1
+                runtimeStats.totalRequestsCountByType[requestType] = (runtimeStats.totalRequestsCountByType[requestType] || 0) + 1
+                runtimeStats.totalRequestsCountByTarget[xTarget] = (runtimeStats.totalRequestsCountByTarget[xTarget] || 0) + 1
+
+                res.on('close', () => {
+                    if (requestType) {
+                        runtimeStats.activeRequestsIds.delete(req.id)
+                        runtimeStats.activeRequestsCountByType[requestType] = Math.max(0, runtimeStats.activeRequestsCountByType[requestType] - 1)
+                        runtimeStats.activeRequestsCountByTarget[xTarget] = Math.max(0, runtimeStats.activeRequestsCountByTarget[xTarget] - 1)
+                    }
+                })
+            } catch (error) {
+                logger.error({
+                    msg: 'runtimeStatsMiddleware error',
+                    error,
+                    data: { url: req.url, method: req.method, runtimeStats },
+                })
+            }
+
+            next()
+        })
+
+        if (conf[ACCESS_TOKEN_VAR_NAME]) {
+            app.get(this.statsUrl, (req, res) => {
+                const token = req.query['token']
+                const reqIds = req.query['reqIds']
+                const accessToken = conf[ACCESS_TOKEN_VAR_NAME]
+                if (!!accessToken && accessToken === token) {
```

Comment:
I guess you should use `crypto.timingSafeEqual` here

Reply:
Done